server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -1834,6 +1846,71 @@ public long getDefaultKeepAliveInMillis() {
         return defaultKeepAlive;
     }
 
+    public List<ReaderContext> getActiveContexts(ShardId shardId) {
+        return this.activeReaders.values()
+            .stream()
+            .filter(c -> c.singleSession() == false)
+            .filter(c -> c.scrollContext() == null)
```

Comment:
exclude single-session ReaderContexts because they get cleared after each search, also don't move anything that is a scroll context.